DEVX-554: Kotlin client example for CCloud

gAmUssA opened the pull request:
kotlin version of helloworld application

miguno commented on kafka-streams/src/test/java/io/confluent/examples/streams/IntegrationTestUtils.java:
```diff
@@ -195,10 +198,77 @@ private static boolean continueConsuming(int messagesConsumed, int maxMessages)
+    long toNowInProcessingTimeMs = System.currentTimeMillis();
+    for (K key : expected.keySet()) {
+      long windowCounter = 0;
+      try (WindowStoreIterator<V> iterator = store.fetch(key, fromBeginningOfTimeMs, toNowInProcessingTimeMs)) {
```
This method operates on the (perhaps incorrect) assumption that `ReadOnlyWindowStore#fetch()` has an ordering guarantee so that the first returned window (in the iterator) is always the oldest available window.  Is this actually true?  We don't specify a contract for `fetch()`.

dguy replied:
For each key, ordering will be guaranteed.

miguno replied:
Is this the official contract?  It's not explicitly specified in the javadocs of `fetch()`.  I'd suggest to update the javadocs (I can send a PR).  WDYT?

Also, do we have the same contract for `ReadOnlySessionStore#fetch()`?

dguy replied:
Yes - we always search from the earliest segment to the latest.

dguy replied:
javadoc update would be helpful. Thanks @miguno

miguno replied:
Kafka docs PR at https://github.com/apache/kafka/pull/2751

miguno commented on kafka-streams/src/test/java/io/confluent/examples/streams/ValidateStateWithInteractiveQueriesLambdaIntegrationTest.java:
```diff
@@ -0,0 +1,149 @@
+        IntegrationTestUtils.waitUntilStoreIsQueryable("max-store", QueryableStoreTypes.keyValueStore(), streams);
+    ReadOnlyWindowStore<String, Long> windowStore =
+        IntegrationTestUtils.waitUntilStoreIsQueryable("max-window-store", QueryableStoreTypes.windowStore(), streams);
+    // Wait one more time so that the input data can be fully processed
```
Is there a better way?  Unfortunately, waiting based on `KafkaStreams#state() == KafkaStreams.State.RUNNING` doesn't work because I have found that the `KafkaStreams` instance may already be in `RUNNING` but the state store is still not yet queryable; that is, the instance is RUNNING but attempting to do `streams.store(...)` results in `InvalidStateStoreException` ("... migrated to a another instance").

dguy replied:
Unfortunately when you call `KafkaStreams.start()` it is immediately in the `RUNNING` state, which is a bit confusing as it isn't really ready. You could loop on:
`KafkaStreams#allMetadataForStore(...)` until it is none empty.

gAmUssA committed:
🚀 initial commit of kotlin version of hello world example

ybyzek commented on clients/cloud/README.md:
```diff
@@ -9,6 +9,8 @@ This directory includes examples of Kafka client applications connecting to [Con
 [![](images/confluent-cli.png)](confluent-cli/)
 [![](images/ruby.png)](ruby/)
 
+[![](images/kotlin.png)](kotlin/)
```
I know you like Kotlin but please resize the image -- it's huge ;)

gAmUssA replied:
Ok

gAmUssA replied:
done

ybyzek commented on clients/cloud/kotlin/build.gradle.kts:
```diff
@@ -0,0 +1,46 @@
+group = "io.confluent.ccloud"
+version = "5.0.1"
```
What does this `version` refer to?

gAmUssA replied:
I copied this from other examples, namely java

ybyzek replied:
ie, will it change with each version of CP?  I don't think so...

mhowlett replied:
cc: @confluentinc/tools 
they get auto-updated in many maven files..

I initially questioned the use of gradle, but it seems preferred by kotlin people.

gAmUssA replied:
will use default (0.0.1-SNAPSHOT)

gAmUssA replied:
@mhowlett i went full blown purist here - gradle build using Kotlin DSL to build kotlin app :)

ybyzek commented on clients/cloud/kotlin/README.adoc:
```diff
@@ -0,0 +1,142 @@
+
+== Prerequisites
+
+* Java 1.8 or higher to run the demo application
```
Is there no Kotlin language-specific prerequisite?

gAmUssA replied:
Not really. Kotlin standard library is in application class path

asm0dey commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/ProducerExample.kt:
```diff
@@ -0,0 +1,100 @@
+
+  } catch (e: InterruptedException) {
+    // Ignore if TopicExistsException, which may be valid if topic exists
+    if (e.cause !is TopicExistsException) {
```
As of me you don't need such complex exception handling here
1. There are no checked exceptions in kotlin
2. Eventually we're catching everything but TopicExistsException.

It looks like we should do smth like

```
catch(e: Exception) {
  if (e.cause !is TopicExistsException) throw e
}
```

gAmUssA replied:
I like that. Thanks

asm0dey commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/ProducerExample.kt:
```diff
@@ -0,0 +1,100 @@
+  // Produce sample data
+  val numMessages = 10L
+  producer.use {
+    for (i in 0L until numMessages) {
```
```
repeat(numMessages){ 
}
```

gAmUssA replied:
This is nice, Thank you

gAmUssA replied:
fixed

asm0dey commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/model/DataRecord.kt:
```diff
@@ -0,0 +1,11 @@
+import com.google.gson.Gson
+
+data class DataRecord(var count: Long) {
+  constructor() : this(0L)
```
I think it will make sense to remove this secondary constructor and add default value to count in primary constructor

gAmUssA replied:
That will work. Thanks

gAmUssA replied:
fixed

gAmUssA committed:
Addressing Yeva’s and Paul’s recommendations

asm0dey commented on clients/cloud/kotlin/build.gradle.kts:
```diff
@@ -0,0 +1,45 @@
+val configPath: String by project
+val topic: String by project
+val mainClass: String by project
+
```
Sorry for posting this late:
I think you need to set jvmTarget to 1.8. Default is 1.6
See here: https://kotlinlang.org/docs/reference/using-gradle.html#attributes-specific-for-jvm

gAmUssA replied:
it's there

```
java {
  sourceCompatibility = JavaVersion.VERSION_1_8
  targetCompatibility = JavaVersion.VERSION_1_8
}
```

asm0dey replied:
That's settings for java, not for kotlin :)

gAmUssA replied:
done.
Thank you

asm0dey commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/util/Utils.kt:
```diff
@@ -0,0 +1,16 @@
+
+fun loadConfig(configFile: String): Properties {
+  if (!Files.exists(Paths.get(configFile))) {
+    throw IOException("$configFile not found.")
```
IllegalArgument?

gAmUssA replied:
?

asm0dey replied:
Is it correct to throw IOException here when there is everything fine with IO?

asm0dey commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/StreamsExample.kt:
```diff
@@ -0,0 +1,88 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:JvmName("StreamsExample")
```
As of me this makes sense only if you use kotlin with java in same project. Otherwise one won't care about filename.

gAmUssA replied:
i use gradle to run my classes. it was throwing `ClassNotFound` without it

asm0dey replied:
But you can call it as StreamsExampleKt

gAmUssA committed:
Small refactoring / simplification

mhowlett commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/ProducerExample.kt:
```diff
@@ -0,0 +1,91 @@
+    with(AdminClient.create(cloudConfig)) {
+      createTopics(listOf(newTopic)).all().get()
+    }
+  } catch (e: Exception) {
```
`ExecutionException`, following the java example, would be marginally better (acts as documentation if nothing else).

side note: the java example shouldn't really catch `InterruptedException` here, since the catch block does nothing with it.

gAmUssA replied:
addressed

mhowlett commented on clients/cloud/kotlin/src/main/kotlin/io/confluent/examples/clients/cloud/ProducerExample.kt:
```diff
@@ -0,0 +1,91 @@
+
+      producer.send(ProducerRecord(topic, key, record)) { m: RecordMetadata,
+                                                          _: Exception? ->
+        println("Produced record to topic ${m.topic()} partition [${m.partition()}] @ offset ${m.offset()}")
```
should you follow the java example here and print out the stack trace in case of error?

gAmUssA replied:
should I? I think I can yes

gAmUssA replied:
addressed

mhowlett commented:
few minor comments, but lgtm from my perspective.

gAmUssA committed:
Adddessing Matt’s comments

gAmUssA commented:
Thanks @mhowlett! Addressed your comments

gAmUssA committed:
Clean up of build script

- addressing Paul’s comments
- simplifying dependencies for Kotlin sodlib and reflect

gAmUssA commented:
Alright, I'm going to merge it then. Thank you all for comments
